sirepo/html.py

```diff
@@ -0,0 +1,62 @@
+# -*- coding: utf-8 -*-
+u"""html template
+
+:copyright: Copyright (c) 2020 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+from __future__ import absolute_import, division, print_function
+from pykern.pkcollections import PKDict
+from pykern.pkdebug import pkdc, pkdlog, pkdp
+import pykern.pkio
+import re
+import sys
+
+
+_CALL_RE = re.compile(r'(\w+)\(([^\)]*)\);', flags=re.MULTILINE)
+
+_ARGS_RE = re.compile(r'\s*,\s*')
+
+_ARG_RE = re.compile(r'^(\w+)=(.*)$', flags=re.DOTALL)
+
+_FUNC_MAP = PKDict()
+
+_FUNC_PREFIX = '_widget_'
+
+def render(path):
+    def _dispatch(match):
+        f = match.group(1)
+        assert f in _FUNC_MAP, \
+            f'function={f} not in map path={path}'
+        k = PKDict(ctx=PKDict(path=path))
+        for a in  _ARGS_RE.split(match.group(2).strip()):
+            m = _ARG_RE.search(a)
+            assert m, \
+                f'invalid keyword argument={a} function={f} path={path}'
+            assert m.group(1) not in k, \
+                f'duplicate keyword argument={a} function={f} path={path}'
+            k[m.group(1)] = m.group(2)
+        return _FUNC_MAP[f](**k)
+
+    return _CALL_RE.sub(_dispatch, pykern.pkio.read_text(path))
+
+
+def _init():
+    import sys, inspect
+    for n, o in inspect.getmembers(sys.modules[__name__]):
+        if n.startswith(_FUNC_PREFIX) and inspect.isfunction(o):
+            _FUNC_MAP[n[len(_FUNC_PREFIX):]] = o
+
+
+def _widget_sim_settings_and_status(ctx, scope):
```

Comment:
I'd rather see the html formatting on the client in javascript. I think we could have a similar html.js module for this. Adding it too the server makes it difficult to switch between python/javascript modes of thinking and make it harder to track down where things are being referenced in the UI.

Reply:
#1561 looms in my mind. How are we ever going to decouple the code from AngularJS so we can migrate to something else? When this has worked in the past, it was through declarative programming: _what not how_.

I realize that we have the problem of switching between python/javascript, and I agree with you. Yet, we have many more modes, and some modes are being ignored such as test mode. When I'm thinking in test mode, I think in code not DOM. 

Javascript in the browser is deficient for test mode in a couple of ways. There's no file system to store intermediate inputs/outputs. This is critical when debugging code generation, because you want to see the output of the code generator. One of the problems with Lisp macros is that all the code generation happens in the interpreter, and you have to do testing at the repl, which seems to be a favorite of lispers, but is insufficient for building large systems. Automation is the only solution, and tracing intermediate results is key to understanding when automation goes wrong.

Another problem with Javascript in the browser is that we lose control of the language processor version (see https://github.com/radiasoft/sirepo/pull/2663#pullrequestreview-455896648). It was very hard with py2/py3, but we could actively choose to be in one or the other. It becomes a least common denominator thing, which to me devolves to assembly language, which is how I think about Javascript. This is why people program in Typescript, in a way, so they can have a consistent execution environment (assuming no bugs in the language processor, which is the C++ problem that they replicated in typescript).

Ideally, we'd build a declarative language that could speak about all the ways we need to interact in Javascript. The schema would drive all of it. Then, we would only need to test the widgets in browsers, independent of the data.

The other thing about generating HTML and Javascript in Javascript is that it is confusing to me to ready. I don't like the templateUrl, because they are not abstracted.

While I understand when we first started writing Sirepo, we didn't know what we are doing. Today, we have lots of examples of what needs to be encapsulated.

For this particular problem, the files are requested by angular directly. Rather than figuring out how angular caches the templates, I decided the server could generate what needs to be cached. This is similar to the google tag solution, and could replace it. The goal would be to start slowly as we modified code, and refine the language over time. I would also like to encaspulate bootstrap as a part of this process, because it will be replaced one day. 

Naturally, this will not make Sirepo easy for the novice to understand. I don't think in its present form it is easy to understand (see [schema constants](https://github.com/radiasoft/sirepo/commit/68d5a2d040d51f8ae5ba41517c9b0571c45a1f72#diff-e804deb9cc0f9aa7d43101b42eb28f55R145)).